DESIGN.md

```diff
@@ -0,0 +1,756 @@
+            }
+            Err(_) => { ... }
+        }
+    }
+}
+```
+
+In practice, operation timeouts will deserve a dedicated API to handle the
+boilerplate.
+
+```rust
+let (tcp_stream, addr) = tcp_listener
+    .accept()
+    .timeout(Duration::from_millis(100))
+    .await?
+```
+
+The cancel and timeout function are inherent methods on operation future types.
+
+If the operation's future drops before the operation completes, the runtime will
+submit a cancellation request for the operation. However, cancellation is
+asynchronous and best-effort, the operation may still complete. In that case,
+the runtime discards the operation result.
+
+The queue's single-producer characteristic optimizes for a single thread to own
+a given submission queue. Supporting a multi-threaded runtime requires either
+synchronizing pushing onto the submission queue or creating a queue pair per
+thread. The tokio-uring runtime uses an isolated thread-per-core model,
+differing from Tokio's other runtimes. Unlike Tokio's primary multi-threaded
+runtime, there is no work-stealing. Tasks remain on the thread that spawned them
+for the duration of their lifecycle. Each runtime thread will own a dedicated
+submission and completion queue pair, and operations are submitted using the
+submission queue associated with the current thread. Operation completion
+futures will not implement `Send`, guaranteeing that they remain on the thread
+to receive the operation result. Interestingly, the resources, e.g.,
+`TcpListener`, can be `Send` as long as they do not hold operation futures
+internally. It is possible to have two operations in-flight from a single
+resource associated with different queues and threads.
+
+Because operations are not `Send`, tasks awaiting these operations are also not
+`Send`, making them unable to be spawned using the spawn function from the Tokio
+crate. The tokio-uring crate will provide a spawn function that accepts not
+`Send` tasks.
+
+When using multiple isolated runtime threads, balancing load between them
+becomes challenging. Applications must take care to ensure load remains balanced
+across threads, and strategies tend to vary. For example, a TCP server can
+distribute accepted connections across multiple threads, ideally while
+maintaining equal load across threads. One approach is to submit "accept"
+operations for the same listener concurrently across all the runtime threads
+while ensuring overloaded workers back-off. Defining runtime load is out of this
+crate's scope and left to the application, though pseudocode follows.
+
+```rust
+let listener = Arc::new(TcpListener::bind(...));
+
+spawn_on_each_thread(async {
+    loop {
+        // Wait for this worker to have capacity. This
+        // ensures there are a minimum number of workers
+        // in the runtime that are flagged as with capacity
+        // to avoid total starvation.
+        current_worker::wait_for_capacity().await;
```

Comment:
Does this imply that "worker" is a new noun in this API? If so, is `spawn_on_each_thread` more like "spawn on each worker?" Or should this be `current_thread::wait_for_capacity` instead?

But also, if the implication is that the only way to saturate your runtime is to always be accepting connections on all threads that are dedicated to it, should `spawn_on_each_thread` take a `FnMut() -> impl Future` instead to remove the `loop`? The `loop` seems error prone, because if it ever exited (due to panic/etc) you'd have an idle thread spawned by the runtime, whereas with a function the runtime could restart threads.

Reply:
It could be `current_thread`, this snippet is just to illustrate the concept and not to include within tokio-uring. Either way, we can explore the options outside of the crate.